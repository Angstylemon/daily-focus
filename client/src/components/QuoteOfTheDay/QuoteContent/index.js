import React from "react";
import PropTypes from "prop-types";
import style from "./style.module.scss";
import { Fab, makeStyles } from "@material-ui/core";
import SettingsIcon from "@material-ui/icons/Settings";

const useStyles = makeStyles({
    btn: {
        background: "#30A0F5",
    },
});

export default function QuoteContent({ text, author, setIsSwitchingSubject }) {
    const classes = useStyles();

    return (
        <div>
            <p className={style.content}>{text}</p>
            <p className={style.author}>{"- " + author}</p>
            <div className={style.settingButtonDiv}>
                <Fab
                    className={style.settingButton}
                    color="primary"
                    size="medium"
                    onClick={() => setIsSwitchingSubject(true)}
                    classes={{ root: classes.btn }}
                >
                    <SettingsIcon />
                </Fab>
            </div>
        </div>
    );
}

QuoteContent.propTypes = {
    text: PropTypes.string.isRequired,
    author: PropTypes.string.isRequired,
    setIsSwitchingSubject: PropTypes.func.isRequired,
};
