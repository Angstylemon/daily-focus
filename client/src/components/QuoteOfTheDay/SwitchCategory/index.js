import PropTypes from "prop-types";
import { Fab } from "@material-ui/core";
import ArrowBackIcon from "@material-ui/icons/ArrowBack";
import RadioButtonCheckedIcon from "@material-ui/icons/RadioButtonChecked";
import RadioButtonUncheckedIcon from "@material-ui/icons/RadioButtonUnchecked";
import style from "./style.module.scss";

function SwitchCategory({ cancelClicked, updateCategory, currentCategory }) {
    return (
        <div className={style.container}>
            <div className={style.listHolder}>
                <div
                    className={style.categoryOption}
                    onClick={() => updateCategory("Taylor Swift")}
                >
                    Taylor Swift
                    {currentCategory === "Taylor Swift" ? (
                        <RadioButtonCheckedIcon />
                    ) : (
                        <RadioButtonUncheckedIcon />
                    )}
                </div>
                <div className={style.categoryOption} onClick={() => updateCategory("Kanye West")}>
                    Kanye West
                    {currentCategory === "Kanye West" ? (
                        <RadioButtonCheckedIcon />
                    ) : (
                        <RadioButtonUncheckedIcon />
                    )}
                </div>
                <div
                    className={style.categoryOption}
                    onClick={() => updateCategory("Chuck Norris")}
                >
                    Chuck Norris
                    {currentCategory === "Chuck Norris" ? (
                        <RadioButtonCheckedIcon />
                    ) : (
                        <RadioButtonUncheckedIcon />
                    )}
                </div>
                <div className={style.categoryOption} onClick={() => updateCategory("Anime")}>
                    Anime ( ͡° ͜ʖ ͡°)
                    {currentCategory === "Anime" ? (
                        <RadioButtonCheckedIcon />
                    ) : (
                        <RadioButtonUncheckedIcon />
                    )}
                </div>
                <div className={style.categoryOption} onClick={() => updateCategory("Programming")}>
                    Programming
                    {currentCategory === "Programming" ? (
                        <RadioButtonCheckedIcon />
                    ) : (
                        <RadioButtonUncheckedIcon />
                    )}
                </div>
            </div>
            <div className={style.settingButtonDiv}>
                <Fab
                    className={style.settingButton}
                    color="primary"
                    size="medium"
                    onClick={() => cancelClicked()}
                >
                    <ArrowBackIcon />
                </Fab>
            </div>
        </div>
    );
}

SwitchCategory.propTypes = {
    cancelClicked: PropTypes.func.isRequired,
    updateCategory: PropTypes.func.isRequired,
    currentCategory: PropTypes.string.isRequired,
};

export default SwitchCategory;
