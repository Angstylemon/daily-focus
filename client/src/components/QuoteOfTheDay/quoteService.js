async function fetchKanyeQuote(setText, setAuthor) {
    await fetch("https://api.kanye.rest")
        .then((res) => {
            res.json().then((jsonRes) => {
                const newQuote = jsonRes.quote;
                if (newQuote) {
                    setText('"' + newQuote + '"');
                    setAuthor("Kanye West");
                    return { text: '"' + newQuote + '"', author: "Kanye West" };
                } else {
                    setText("Sorry, could not get quote");
                    setAuthor("");
                }
            });
        })
        .catch(() => {
            setText("Sorry, could not get quote");
            setAuthor("");
        });
}

async function fetchChuckNorrisQuote(setText, setAuthor) {
    await fetch("https://api.chucknorris.io/jokes/random")
        .then((res) => {
            res.json().then((jsonRes) => {
                const newQuote = jsonRes.value;
                if (newQuote) {
                    setText('"' + newQuote + '"');
                    setAuthor("Anonymous");
                    return { text: '"' + newQuote + '"', author: "Anonymous" };
                } else {
                    setText("Sorry, could not get quote");
                    setAuthor("");
                }
            });
        })
        .catch(() => {
            setText("Sorry, could not get quote");
            setAuthor("");
        });
}

async function fetchProgrammingQuote(setText, setAuthor) {
    await fetch("http://quotes.stormconsultancy.co.uk/random.json")
        .then((res) => {
            res.json().then((jsonRes) => {
                const newQuote = jsonRes.quote;
                if (newQuote) {
                    setText('"' + newQuote + '"');
                    setAuthor(jsonRes.author);
                    return { text: '"' + newQuote + '"', author: jsonRes.author };
                } else {
                    setText("Sorry, could not get quote");
                    setAuthor("");
                }
            });
        })
        .catch(() => {
            setText("Sorry, could not get quote");
            setAuthor("");
        });
}

async function fetchTaylorSwiftQuote(setText, setAuthor) {
    await fetch("https://api.taylor.rest/")
        .then((res) => {
            res.json().then((jsonRes) => {
                const newQuote = jsonRes.quote;
                if (newQuote) {
                    setText('"' + newQuote + '"');
                    setAuthor("Taylor Swift");
                    return { text: '"' + newQuote + '"', author: "Taylor Swift" };
                } else {
                    setText("Sorry, could not get quote");
                    setAuthor("");
                }
            });
        })
        .catch(() => {
            setText("Sorry, could not get quote");
            setAuthor("");
        });
}

async function fetchAnimeQuote(setText, setAuthor) {
    await fetch("https://animechan.vercel.app/api/random")
        .then((res) => {
            res.json().then((jsonRes) => {
                const newQuote = jsonRes.quote;
                if (newQuote) {
                    setText('"' + newQuote + '"');
                    setAuthor(jsonRes.character + ", " + jsonRes.anime);
                    return {
                        text: '"' + newQuote + '"',
                        author: jsonRes.character + ", " + jsonRes.anime,
                    };
                } else {
                    setText("Sorry, could not get quote");
                    setAuthor("");
                }
            });
        })
        .catch(() => {
            setText("Sorry, could not get quote");
            setAuthor("");
        });
}

export {
    fetchKanyeQuote,
    fetchChuckNorrisQuote,
    fetchTaylorSwiftQuote,
    fetchProgrammingQuote,
    fetchAnimeQuote,
};
