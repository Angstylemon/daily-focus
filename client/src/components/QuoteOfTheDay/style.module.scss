.container {
    display: flex;
    flex-direction: column;
    width: 340px;
    height: fit-content;
    background-color: #ffffff;
    border-radius: 25px;
    box-shadow: 0px 4px 16px 8px rgba(0, 0, 0, 0.1);

    .title {
        display: flex;
        justify-content: center;
        align-items: center;
    }

    .title {
        display: flex;
        align-items: center;
        justify-content: center;
        background: #30a0f5;
        height: 48px;
        width: 100%;
        border-top-left-radius: 25px;
        border-top-right-radius: 25px;

        &__text {
            color: white;
            font-size: 24px;
            font-family: "Ropa Sans";
        }
    }

    .content {
        display: flex;
        justify-content: center;
        align-items: center;
        padding: 10px;
        text-align: center;
    }

    .author {
        font-size: 16px;
        font-style: italic;
        text-align: center;
    }

    .settingButtonDiv {
        display: flex;
        justify-content: flex-end;
        align-items: flex-end;
    }
    .settingButton {
        background-color: #30a0f5;
        margin-right: 10px;
        margin-bottom: 10px;
    }
}
