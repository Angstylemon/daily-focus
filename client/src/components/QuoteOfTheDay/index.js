import React, { useEffect, useState } from "react";
import createPersistedState from "use-persisted-state";
import { CircularProgress } from "@material-ui/core";
import style from "./style.module.scss";
import SwitchCategory from "./SwitchCategory";
import QuoteContent from "./QuoteContent";
import {
    fetchKanyeQuote,
    fetchInspirationalQuote,
    fetchTaylorSwiftQuote,
    fetchProgrammingQuote,
    fetchAnimeQuote,
} from "./quoteService";

const quoteCategories = [
    {
        value: "inspirational",
        text: "Inspirational",
        fetchQuote: (textHandler, authorHandler) =>
            fetchInspirationalQuote(textHandler, authorHandler),
    },
    {
        value: "taylor-swift",
        text: "Taylor Swift",
        fetchQuote: (textHandler, authorHandler) =>
            fetchTaylorSwiftQuote(textHandler, authorHandler),
    },
    {
        value: "kanye-west",
        text: "Kanye West",
        fetchQuote: (textHandler, authorHandler) => fetchKanyeQuote(textHandler, authorHandler),
    },
    {
        value: "programming",
        text: "Programming",
        fetchQuote: (textHandler, authorHandler) =>
            fetchProgrammingQuote(textHandler, authorHandler),
    },
    {
        value: "anime",
        text: "Anime ( ͡° ͜ʖ ͡°)",
        fetchQuote: (textHandler, authorHandler) => fetchAnimeQuote(textHandler, authorHandler),
    },
];

const useQuoteCategoryState = createPersistedState("quoteCategory");

export default function QuoteOfTheDay() {
    const [isLoading, setIsLoading] = useState(true);
    const [isSwitchingSubject, setIsSwitchingSubject] = useState(false);
    const [category, setCategory] = useQuoteCategoryState("inspirational");
    const [text, setText] = useState("");
    const [author, setAuthor] = useState("");

    useEffect(() => {
        quoteCategories.forEach(async (cat) => {
            if (cat.value === category) {
                await cat.fetchQuote(setText, setAuthor);
                setIsLoading(false);
            }
        });
    }, [category]);

    /**
     * Callback for when a new category is selected
     * @param value The new category
     */
    function updateContent(value) {
        setIsLoading(true);
        setCategory(value);
    }

    return (
        <div className={style.container}>
            <div className={style.title}>
                <span className={style.title__text}>Quote of the day</span>
            </div>
            {isSwitchingSubject ? (
                <SwitchCategory
                    categories={quoteCategories}
                    cancelClicked={() => setIsSwitchingSubject(false)}
                    updateCategory={updateContent}
                    currentCategory={category}
                />
            ) : isLoading ? (
                <div className={style.loading}>
                    <CircularProgress />
                </div>
            ) : (
                <QuoteContent
                    text={text}
                    author={author}
                    setIsSwitchingSubject={setIsSwitchingSubject}
                />
            )}
        </div>
    );
}
