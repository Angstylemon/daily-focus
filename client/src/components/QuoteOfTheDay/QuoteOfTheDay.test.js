import React, { useState } from "react";
import { shallow } from "enzyme";
import ShallowRenderer from "react-test-renderer/shallow";
import {
    fetchKanyeQuote,
    fetchChuckNorrisQuote,
    fetchTaylorSwiftQuote,
    fetchProgrammingQuote,
    fetchAnimeQuote,
} from "./quoteService";

import QuoteOfTheDay from ".";

let component;

function setText(text) {
    console.log(text);
}

function setAuthor(author) {
    console.log(author);
}

beforeEach(() => {
    component = shallow(<QuoteOfTheDay />);
});

test("it renders correctly", () => {
    const shallowRenderer = new ShallowRenderer();
    const snapshotComponent = shallowRenderer.render(<QuoteOfTheDay />);
    expect(snapshotComponent).toMatchSnapshot();
});

test("It fetches Kanye West quotes correctly", () => {
    let { text, author } = fetchKanyeQuote(setText, setAuthor);
    expect(text !== "");
    expect(author === "Kanye West");
});

test("It fetches Chuck Norris quotes correctly", () => {
    let { text, author } = fetchChuckNorrisQuote(setText, setAuthor);
    expect(text !== "");
    expect(author === "Anonymous");
});

test("It fetches Taylor Swift quotes correctly", () => {
    let { text, author } = fetchTaylorSwiftQuote(setText, setAuthor);
    expect(text !== "");
    expect(author === "Taylor Swift");
});

test("It fetches programming quotes correctly", () => {
    let { text, author } = fetchProgrammingQuote(setText, setAuthor);
    expect(text !== "");
    expect(author !== "");
});

test("It fetches anime quotes correctly", () => {
    let { text, author } = fetchAnimeQuote(setText, setAuthor);
    expect(text !== "");
    expect(author !== "");
});
