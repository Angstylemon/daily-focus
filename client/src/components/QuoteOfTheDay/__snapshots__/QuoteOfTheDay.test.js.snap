// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders correctly 1`] = `
<div
  className="container"
>
  <div
    className="title"
  >
    <span
      className="title__text"
    >
      Quote of the day
    </span>
  </div>
  <div
    className="loading"
  >
    <WithStyles(ForwardRef(CircularProgress)) />
  </div>
</div>
`;
